node: device-mgr: Handle recovery flow by checking if healthy devices exist- attempt 2

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
This is an attempt to merge changes proposed as part of https://github.com/kubernetes/kubernetes/pull/114640. The PR was [reverted](https://github.com/kubernetes/kubernetes/pull/116341) due to serial test [failures](https://github.com/kubernetes/kubernetes/issues/116340) and this PR updates device plugin and device manager e2e tests to avoid serial test failures.

Core changes were squshed into a single commit: [dc1a592632](https://github.com/kubernetes/kubernetes/pull/116376/commits/dc1a5926321410f141f48b9edd3397fed2d0f93e) `node: device-mgr: Handle recovery by checking if healthy devices exist` which was already approved in #114640 (https://github.com/kubernetes/kubernetes/pull/116376/commits/0aa6a5726a67edd4ffc0efd1af7179cf47e532ce, https://github.com/kubernetes/kubernetes/pull/114640/commits/a799ffb571476d7fb706009e79aaf933b7851991 and https://github.com/kubernetes/kubernetes/pull/114640/commits/5b2a3dbbdc9cd25c09b65dd8c4ca630b001ab084).

Other commits are related to device plugin and device manager e2e tests. The PR adds a bunch of test cases to ensure comprehensive testing but the main focus is to recreate a scenario (where on kubelet restart application pod requesting devices comes up before device plugin pod re-registers itself) resulting in pod admission error.

**Rationale of the PR as captured in the previous PR:**

In case of node reboot/kubelet restart, the flow of events involves obtaining the state from the checkpoint file followed by setting the `healthDevices`/`unhealthyDevices` to its zero value. This is done to allow the device plugin to re-register itself so that capacity can be updated appropriately.

During the allocation phase, we need to check if the resources requested by the pod have been registered AND healthy devices are present on the node to be allocated.

Also we need to move this check above `needed==0` where needed is required - devices allocated to the container (which is obtained from the checkpoint file) because even in cases where no additional devices have to be allocated (as they were pre-allocated), we still need to make sure that the devices that were previously allocated are healthy.

For more details refer to the comment here: https://github.com/kubernetes/kubernetes/issues/109595#issuecomment-1359882423.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #109595 and (possibly) #107928 (TODO: need to check the latter)
#### Special notes for your reviewer:

**NOTE:** 

1. **This PR depends on https://github.com/kubernetes/kubernetes/pull/117057 for device plugin e2e test refactoring work and builds on top of that.** 

2. **This PR depends on sample device plugin changes proposed in: https://github.com/kubernetes/kubernetes/pull/115107.** 

Those changes are included here (first two commits) as they are needed for e2e test implementation.

NOTE: On node reboot, in case the application pod requesting devices starts after the device plugin pod providing those devices as has been described in https://github.com/kubernetes/kubernetes/issues/109595 and https://github.com/kubernetes/kubernetes/issues/107928 we exit early as there are pre-allocated devices without checking if the devices have registered and are healthy. This PR adds both those checks.

Because of this change, in cases where application pods are started before device plugin pod (say after node reboot),  because devices are not healthy, the pod would fail with `UnexpectedAdmissionError` error.  If the pod is part of a deployment, another pod would be created but that would stay in `Pending` state waiting for the pod to be scheduled to the node. This pod would go `Running` after the node capacity is updated followed by start up of device plugin pod and its registration.

The pod which fails at admission time continues to exist on the node and needs to be removed manually:
```
kubectl delete pods --field-selector status.phase=Failed
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
During device plugin allocation, resources requested by the pod can only be allocated if the device plugin has registered itself to kubelet AND healthy devices are present on the node to be allocated. If these conditions are not sattsfied, the pod would fail with `UnexpectedAdmissionError` error.
```
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

swatisehgal committed:
node: device-mgr: Handle recovery by checking if healthy devices exist

In case of node reboot/kubelet restart, the flow of events involves
obtaining the state from the checkpoint file followed by setting
the `healthDevices`/`unhealthyDevices` to its zero value. This is
done to allow the device plugin to re-register itself so that
capacity can be updated appropriately.

During the allocation phase, we need to check if the resources requested
by the pod have been registered AND healthy devices are present on
the node to be allocated.

Also we need to move this check above `needed==0` where needed is
required - devices allocated to the container (which is obtained from
the checkpoint file) because even in cases where no additional devices
have to be allocated (as they were pre-allocated), we still need to
make sure he devices that were previously allocated are healthy.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: e2e: Implement End to end test

This commit reuses e2e tests implmented as part of https://github.com/kubernetes/kubernetes/pull/110729.
The commit is borrowed from the aforementioned PR as is to preserve
authorship. Subsequent commit will update the end to end test to
simulate the problem this PR is trying to solve by reproducing
the issue: 109595.

Co-authored-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: e2e: Update the e2e test to reproduce issue:109595

Breakdown of the steps implemented as part of this e2e test is as follows:
1. Create a file `registration` at path `/var/lib/kubelet/device-plugins/sample/`
2. Create sample device plugin with an environment variable with
   `REGISTER_CONTROL_FILE=/var/lib/kubelet/device-plugins/sample/registration` that
    waits for a client to delete the control file.
3. Trigger plugin registeration by deleting the abovementioned directory.
4. Create a test pod requesting devices exposed by the device plugin.
5. Stop kubelet.
6. Remove pods using CRI to ensure new pods are created after kubelet restart.
7. Restart kubelet.
8. Wait for the sample device plugin pod to be running. In this case,
   the registration is not triggered.
9. Ensure that resource capacity/allocatable exported by the device plugin is zero.
10. The test pod should fail with `UnexpectedAdmissionError`
11. Delete the test pod.
12. Delete the sample device plugin pod.
13. Remove `/var/lib/kubelet/device-plugins/sample/` and its content, the directory
    created to control registration

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: e2e: adapt to sample device plugin refactoring

These updates are to adapt to the sample device plugin
refactoring done here: https://github.com/kubernetes/kubernetes/pull/115926/commits/92e00203e027bf88ddc695fc17e80c913a6b178a.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-unit

swatisehgal commented:
/cc @SergeyKanzhelev @ffromani @bobbypage @pohly @bart0sh @klueska PTAL

bart0sh commented:
/triage accepted
/priority important-longterm

bart0sh commented:
/assign @ffromani @klueska

ffromani commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -227,21 +227,13 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
-			restartKubelet(true)
-
 			ginkgo.By("Wait for node to be ready again")
 			e2enode.WaitForAllNodesSchedulable(ctx, f.ClientSet, 5*time.Minute)
```
without kubelet restart, these two lines seem redundant, surely the `gingko.By`. We may want to keep the `waitForallNodesSchedulable` for extra safety though

ffromani commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -217,7 +217,7 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
 		})
 
-		ginkgo.It("Keeps device plugin assignments across pod and kubelet restarts", func(ctx context.Context) {
+		ginkgo.It("Keeps device plugin assignments across pod restarts (no kubelet restart)", func(ctx context.Context) {
```
I think it's a good idea to split the previous testcase which was testing for BOTH kubelet restart AND container (pod?) restart. We should have two `It()` blocks each corresponding to the cases, and later we should check if/how combine them.

swatisehgal replied:
Sure, I can do that.

ffromani reviewed (commented):
initial review
kudos for quickly addressing the test failures. I agree that device_plugin_test can use some improvements/updates

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -238,7 +238,7 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
 		})
 
-		ginkgo.It("Keeps device plugin assignments after the device plugin has been re-registered", func(ctx context.Context) {
+		ginkgo.It("Keeps device plugin assignments across kubelet restart after the device plugin has been re-registered", func(ctx context.Context) {
```
So in summary you want me to have the following test cases:

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -280,6 +284,8 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
 
 			ginkgo.By("Creating another pod")
 			pod2 := e2epod.NewPodClient(f).CreateSync(ctx, makeBusyboxPod(SampleDeviceResourceName, podRECMD))
+			err = e2epod.WaitTimeoutForPodRunningInNamespace(ctx, f.ClientSet, pod2.Name, f.Namespace.Name, 1*time.Minute)
```
Sure, I will try to do a better job by adding more context and expressing the rationale behind changes the relevant commit messages.

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -264,9 +264,13 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+			framework.ExpectNoError(err)
+
+			ginkgo.By("Confirming that after a kubelet restart, fake-device assignment is kept")
+			err = e2epod.WaitTimeoutForPodRunningInNamespace(ctx, f.ClientSet, pod1.Name, f.Namespace.Name, 1*time.Minute)
```
I see what you mean but my intention here was to capture the fact that after kubelet restarted and the device plugin re-registered we confirm the device assignment. Admittedly, the statement in the `By` clause does not reflect that accurately. The check is not immediately after kubelet restart but it is still after kubelet restart :) Perhaps it would be better to update the statement or move the check right after kubelet restart.
I will make update to address this.

ffromani commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -236,9 +236,133 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+			framework.ExpectEqual(devIDRestart1, devID1)
+		})
+
+		ginkgo.It("Keeps device plugin assignments across pod and kubelet restarts (no device plugin re-registration)", func(ctx context.Context) {
```
this is a nice extra!

ffromani commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -236,9 +236,133 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+		})
+
+		ginkgo.It("Keeps device plugin assignments after the device plugin has been re-registered (no kubelet or pod restart)", func(ctx context.Context) {
+			podRECMD := "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo stub devices: $devs && sleep 60"
```
do we want to increase the sleep period here, effectively making "forever" for the test timescale?

swatisehgal replied:
Sure, we should definitely do that.

ffromani reviewed (commented):
thanks for the updates. This is definitely moving in the right direction

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -236,9 +236,133 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+			ginkgo.By("Checking that pod got a fake device")
+			devID2 := parseLog(ctx, f, pod2.Name, pod2.Name, deviceIDRE)
+
+			gomega.Expect(devID2).To(gomega.Not(gomega.Equal("")))
```
Ack, sure. Will update.

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -217,7 +217,7 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
 		})
 
-		ginkgo.It("Keeps device plugin assignments across pod restarts (no kubelet restart)", func(ctx context.Context) {
+		ginkgo.It("Keeps device plugin assignments across pod restarts (no kubelet restart and device plugin re-registration)", func(ctx context.Context) {
```
Sure, will update.

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -236,9 +236,133 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
+		})
+
+		ginkgo.It("Keeps device plugin assignments across kubelet restarts (no pod restart and device plugin re-registration)", func(ctx context.Context) {
+			podRECMD := "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo stub devices: $devs && sleep 60"
```
Yeah, good point. For test cases where we don't want pod restart we should change the sleep to "forever"? Let's agree on what is a suitable value here. I am thinking an hour should be reasonable here. WDYT?

ffromani replied:
This has been a significant source of confusion for me, so yes, I believe tests which not involve the pod restart should sleep "forever".  To better convey this intent I'd use some huge value like `24h`, but indeed anything >= `1h` seems OK. Your call if you want to add a constant in the test code to convey the intent or not.

swatisehgal replied:
Ack, let's use a big value like `24h` to emphasize this idea. I will use a constant and provide a comment to express the intent .

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

ffromani commented:
`waitForNodeReady` has to be reimplemented/replaced becayse it's linux-specific

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-e2e-kind

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/assign @klueska 
for review/approval

The core changes and device manager e2e tests were cherry-picked from https://github.com/kubernetes/kubernetes/pull/114640. The changes in the PR are related to device plugin e2e test.

ffromani commented:
thanks @swatisehgal for your continued work on this bug. I'll totally have a look today.

ffromani commented:
> Checking the new changes versus the content of the old PR, almost everything I see here makes sense. Sure, in some areas the tests can be further improved, but the existing changes are improvements already and we'd better iterate than fixing everything in one go.
> 
> The "almost" is: it's still is not clear to me is what guarantees the app pod is always recovered before the device plugin pod. The e2e tests makes fully sense in this scenario, but is not obvious why this is always the case.
> 
> I'll have another pass later today.

The ordering in the recovering flow when kubelet is restarted is my main and only concern here. We can improve later other points.

ffromani commented:
/kind bug

swatisehgal committed:
node: device-plugin: e2e: Capture pod admission failure

This test captures that scenario where after kubelet restart,
application pod comes up and the device plugin pod hasn't re-registered
itself, the pod fails with admission error. It is worth noting that
once the device plugin pod has registered itself, another
application pod requesting devices ends up running
successfully.

For the test case where kubelet is restarted and device plugin
has re-registered without involving pod restart, since the
pod after kubelet restart ends up with admission error,
we cannot be certain the device that the second pod (pod2) would
get. As long as, it gets a device we consider the test to pass.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-plugin: add node reboot test scenario

Add a test suit to simulate node reboot (achieved by removing pods
using CRI API before kubelet is restarted).

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

swatisehgal commented:
/retest

swatisehgal committed:
node: device-plugin: e2e: Additional test cases

Additional test cases added:
Keeps device plugin assignments across pod and kubelet restarts (no device plugin re-registration)
Keeps device plugin assignments after the device plugin has re-registered (no kubelet or pod restart)

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

rphillips commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -544,16 +544,30 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 
+	// Check if registered resource has healthy devices
+	if healthyDevices.Len() == 0 {
+		return nil, fmt.Errorf("can't allocate unhealthy devices %s", resource)
```
This error message doesn't look correct for a zero length check.

swatisehgal replied:
This check is to ensure that we rule out cases where we don't have healthy devices that are suitable for allocation, hence the message. Would you prefer it to be more descriptive such as  `fmt.Errorf("no healthy devices present, can't allocate unhealthy devices %s", resource)`?

rphillips replied:
I think that is very helpful!

swatisehgal replied:
Updated with a more descriptive error message.

rphillips commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +275,239 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			  waits for a client to delete the control file.
+		   3. Trigger plugin registeration by deleting the abovementioned directory.
+		   4. Create a test pod requesting devices exposed by the device plugin.
+		   5. Stop kubelet.
```
IIRC, if we stop the kubelet then we have to label the test with [Disruptive].

swatisehgal replied:
Good point, will add that. Thanks!

swatisehgal replied:
Updated.

swatisehgal commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +275,239 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+
+				defer ginkgo.GinkgoRecover()
+				framework.Logf("Deleting the control file: %q to trigger registration", triggerPathFile)
+				err := os.Remove(triggerPathFile)
```
We are using a modified version of sample-device plugin for testing here. https://github.com/kubernetes/kubernetes/pull/115107. In this version we control the registration of the device plugin with an environment variable `REGISTER_CONTROL_FILE` . The deletion of the file triggers registration of the device plugin to Kubelet.  

We implemented this in a go routine because the first time the device plugin pod is deployed, we want it to register to  kubelet as normal. However, after kubelet restart we want to artificially create a scenario where application pods appears before device plugin pod by not letting the plugin re-register to kubelet.

rphillips replied:
Thanks for that clarification!

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

ffromani commented:
/retest

failures are unrelated

swatisehgal commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e

ffromani commented:
/test pull-kubernetes-e2e-kind-ipv6

unrelated sig-apps failures

rphillips commented:
/lgtm
/assign @mrunalp

ffromani reviewed (commented):
/lgtm
